### ISSUE TITLE
Optimization of Force invalidation

### ISSUE THREAD

**Issue by jchodera** [openmm/openmm#1981]
Currently, changing a `Context` parameter will invalidate all forces, even if the parameter only affects a single `Custom*Force` that is inexpensive to recompute (such as a `CustomCompoundForce` where the center of an umbrella restraint is being moved). 

If, instead, track force validity by individual `Force` object, some operations---such as changing a `Context` parameter---would only invalidate one or a few forces, allowing us to generally avoid recomputing energies or forces for those that don't require it.

This potentially simple optimization could provide enormous speedups for a number of our use cases:
* Alchemical free energy calculations, which frequently requires we sweep over 50-100 different values of a context parameter to compute energies (via `Context.getState()`) in which only some `Custom*Force`s are affected
* Umbrella sampling, which requires the same operation except only some restraint forces are affected
* Nonequilibrium switching, where a `CustomIntegrator` switches a context parameter during integration to accumulate work from only a few `Custom*Force`s

Additionally, we're not sure if changing a context parameter triggers neighborlist regeneration, but we can't think of a case where it would ever need to since context parameters cannot affect the cutoff.

cc: https://github.com/pandegroup/openmm/issues/1980

**Comment by peastman** [comment]
Can you provide an example that demonstrates the problem?  I looked through the code and everything looks correct to me.  Each force only checks to see if its own parameters have changed.  It doesn't check the ones defined by other forces.

**Comment by jchodera** [comment]
Sure! Here's a simple test system where we load a solvated kinase (55K atoms) and add a `CustomBondForce` with a global parameter that controls an umbrella restraint equilibrium distance. After running a step of dynamics to compile kernels and engage constraints, we compute the time required to sweep through the global context parameter (that only affects `CustomBondForce`). We then compute the time required *just* to compute the `CustomBondForce` contribution in a separate force group.

Here's the timing on my laptop:
```
lski2414:timing-comparison choderaj$ python timing-comparison.py 
Loading system...
System has 55541 atoms.
Creating Context...
Timing parameter sweep computing energy for all forces...
   Sweep of umbrella_r0 through 114 took 9.259 s (81.221 ms / energy evaluation)
Timing parameter sweep computing energy for just CustomBondForce...
   Sweep of umbrella_r0 through 114 took 3.628 s (31.829 ms / energy evaluation)
```
Here's the timing on a GTX-1080:
```
[chodera@lilac:timing-comparison]$ python timing-comparison.py 
Loading system...
System has 55541 atoms.
Creating Context...
Timing parameter sweep computing energy for all forces...
   Sweep of umbrella_r0 through 114 took 0.584 s (5.122 ms / energy evaluation)
Timing parameter sweep computing energy for just CustomBondForce...
   Sweep of umbrella_r0 through 114 took 0.069 s (0.605 ms / energy evaluation)
```
[timing-comparison.zip](https://github.com/pandegroup/openmm/files/1693702/timing-comparison.zip)

**Comment by jchodera** [comment]
All the files needed to reproduce this should be in the included ZIP file above.

As you can see, the speed difference is nearly 10x on the GTX-1080.

**Comment by jchodera** [comment]
cc: @andrrizzi

**Comment by jchodera** [comment]
I think I see where the confusion is! I think you're right that `CustomIntegrator` has logic [here](https://github.com/pandegroup/openmm/blob/66bc28f5e91da32744ba076eb2666a7d9164abe6/openmmapi/src/CustomIntegratorUtilities.cpp#L107-L116) to determine which force terms need to be updated by parameter changes, but `Context.getState()` doesn't yet have this kind of optimization built in.

If so, that would mean bullet point 3 (nonequilibrium switching) is already taken care of by the `CustomIntegrator` logic [here](https://github.com/pandegroup/openmm/blob/66bc28f5e91da32744ba076eb2666a7d9164abe6/openmmapi/src/CustomIntegratorUtilities.cpp#L107-L116), but that the first two bullet points could use optimization.

**Comment by jchodera** [comment]
> Context doesn't make any effort to remember the results of past calls to getState(). Every time you call it, it does the calculation all over again. 

That's why this feature request suggests that we change that behavior to individually invalidate forces. This would transparently speed up energy and force evaluation via `Context.getState()` without the potential for causing incorrect results, and without the inconvenience of needing to carefully dissect the system into separate force groups (which is sometimes impossible, or where it is often more useful to use force groups for other purposes, such as MTS integration).

In other words, you're saying "OpenMM doesn't do this now", which is why it's a feature request. :)

**Comment by jchodera** [comment]
Also, apologies for contributing to this confusion---I hadn't realized `Context.getState()` would *always* recompute the forces and did no caching. I had noted that [`ForceImpl` tracked a `forcesInvalid` variable](https://github.com/pandegroup/openmm/blob/15b9fb4894415614514f89e5194e3ee39e8b0ecd/openmmapi/src/ForceImpl.cpp#L37-L42) and had presumed this meant there was some kind of caching going on.

**Comment by peastman** [comment]
It's not a matter of invalidating something.  There simply isn't anything to invalidate.  The context doesn't get a listing of forces and energies broken down by Force object, or even by force group.  It just gets the total force and total energy.  There's a single set of buffers on the GPU for holding forces and energies, and anything that's included in the current computation gets added to those buffers.  A single kernel may even compute several different Forces that are in different groups.  It just accumulates all of them together and saves the total to the buffers.  That total is the only thing the Context ever sees.  It doesn't know what part of it came from each force.

I don't think changing this behavior is a viable option.  It would have major consequences for performance, since we'd need to have a whole lot of different buffers, clear all of them at the start of a step, accumulate every force independently, then reduce each of the buffers separately and send all of them to the CPU.

**Comment by jchodera** [comment]
I see what you mean.

Let me think about this for a bit. We can generally optimize this *manually* by splitting forces into separate force groups and being clever about which groups we trigger energy computation for during `Context.getState()`, and in `CutomIntegrator` steps we can compute the energy or forces for specific force groups. So the capability to carry out the major use cases above is there---just not fully general. I just wonder if there's a way to be more clever automatically, rather than forcing the user to go through the cumbersome practice of carefully splitting forces and computing just the components that are changing.
